v2/apiserver/internal/core/mongodb/events_store.go

```diff
@@ -446,22 +457,32 @@ func (e *eventsStore) DeleteMany(
 	)
 	cur, err := e.collection.Find(ctx, criteria, findOptions)
 	if err != nil {
-		return events, errors.Wrapf(
+		return nil, errors.Wrapf(
 			err,
 			"error finding logically deleted events",
 		)
 	}
-	if err := cur.All(ctx, &events.Items); err != nil {
-		return events, errors.Wrap(
-			err,
-			"error decoding logically deleted events",
-		)
-	}
+
+	eventCh := make(chan core.Event)
+	go func() {
+		defer close(eventCh)
+		for cur.Next(ctx) {
+			event := core.Event{}
+			if err := cur.Decode(&event); err != nil {
+				log.Println(errors.Wrap(err, "unable to decode event"))
+			}
+			select {
+			case eventCh <- event:
+			case <-ctx.Done():
+				return
+			}
+		}
+	}()
 
 	// Final deletion
 	if _, err := e.collection.DeleteMany(ctx, criteria); err != nil {
-		return events, errors.Wrap(err, "error deleting events")
+		return eventCh, errors.Wrap(err, "error deleting events")
 	}
```

Comment:
This final deletion can happen while we're still iterating (goroutine above) over the things that were _slated_ for deletion. Those two concerns are going to race with each other and yield funky results.